cmd/kubeapps-apis/server/packages.go

```diff
@@ -118,6 +123,139 @@ func (s packagesServer) GetAvailablePackageDetail(ctx context.Context, request *
 	}, nil
 }
 
+// GetInstalledPackageSummaries returns the installed package summaries based on the request.
+func (s packagesServer) GetInstalledPackageSummaries(ctx context.Context, request *packages.GetInstalledPackageSummariesRequest) (*packages.GetInstalledPackageSummariesResponse, error) {
+	contextMsg := ""
+	if request.Context != nil {
+		contextMsg = fmt.Sprintf("(cluster=[%s], namespace=[%s])", request.Context.Cluster, request.Context.Namespace)
+	}
+
+	log.Infof("+core GetInstalledPackageSummaries %s", contextMsg)
+
+	// Aggregate the response for each plugin
+	pkgs := []*packages.InstalledPackageSummary{}
+	// TODO: We can do these in parallel in separate go routines.
+	for _, p := range s.plugins {
+		response, err := p.server.GetInstalledPackageSummaries(ctx, request)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Invalid GetInstalledPackageSummaries response from the plugin %v: %v", p.plugin.Name, err)
+		}
+
+		// Add the plugin for the pkgs
+		pluginPkgs := response.InstalledPackageSummaries
+		for _, r := range pluginPkgs {
+			if r.InstalledPackageRef == nil {
+				r.InstalledPackageRef = &packages.InstalledPackageReference{}
+			}
+			r.InstalledPackageRef.Plugin = p.plugin
+		}
+		pkgs = append(pkgs, pluginPkgs...)
+	}
+
+	// Build the response
+	// TODO: Sort via default sort order or that specified in request.
+	return &packages.GetInstalledPackageSummariesResponse{
+		InstalledPackageSummaries: pkgs,
+	}, nil
+}
+
+// GetInstalledPackageDetail returns the package versions based on the request.
+func (s packagesServer) GetInstalledPackageDetail(ctx context.Context, request *packages.GetInstalledPackageDetailRequest) (*packages.GetInstalledPackageDetailResponse, error) {
+	contextMsg := ""
+	if request.InstalledPackageRef != nil && request.InstalledPackageRef.Context != nil {
+		contextMsg = fmt.Sprintf("(cluster=[%s], namespace=[%s])", request.InstalledPackageRef.Context.Cluster, request.InstalledPackageRef.Context.Namespace)
+	}
+
+	log.Infof("+core GetInstalledPackageDetail %s", contextMsg)
+
+	// Check prerequsites
```

Comment:
I believe, in general, it's better to have an early prerequisites-checking step to prevent unhandled situations on the other side. Given that our plugins' APIs are intended to also be called directly, it's true perhaps we are duplicating checking efforts.
So, yep, I totally agree with you, we should either remove it or extract it to a public function.

Besides, a follow-up refactor thing I wanna do at some point is generalizing the logic of each core operation. Right now there are just two scenarios: fetchAll (eg, getXXSummaries) and lookup (eg, getXXDetails). However, I'd rather wait until we have also the actions implemented.